File: tools/tracer/src/tracer/extraction/python.py
"""Tree-sitter extraction for Python.

Extracts module-level imports (both `import x` and `from x import y` forms)
and module-level definitions (functions, classes, top-level constants).
Nested definitions and method definitions are intentionally excluded —
the architecture graph models module-public surface, not internal structure.
"""

from __future__ import annotations

from functools import lru_cache

from tracer.extraction.dispatch import Export, ExtractionResult, Import


@lru_cache(maxsize=1)
def _parser():
    import tree_sitter_python
    from tree_sitter import Language, Parser

    return Parser(Language(tree_sitter_python.language()))


@lru_cache(maxsize=1)
def _query():
    import tree_sitter_python
    from tree_sitter import Language, Query

    language = Language(tree_sitter_python.language())
    return Query(
        language,
        """
        ; import_statement covers `import a`, `import a.b`, `import a as c`
        (import_statement
          name: (dotted_name) @import.module)
        (import_statement
          name: (aliased_import
                  name: (dotted_name) @import.module))

        ; import_from_statement covers `from x import y`, `from x import y as z`
        (import_from_statement
          module_name: (dotted_name) @import_from.module
          name: (dotted_name) @import_from.symbol)
        (import_from_statement
          module_name: (dotted_name) @import_from.module
          name: (aliased_import
                  name: (dotted_name) @import_from.symbol))

        ; module-level definitions only — nested defs are filtered below
        (module
          (function_definition name: (identifier) @export.function))
        (module
          (class_definition name: (identifier) @export.class))
        (module
          (decorated_definition
            definition: (function_definition name: (identifier) @export.function)))
        (module
          (decorated_definition
            definition: (class_definition name: (identifier) @export.class)))
        """,
    )


def extract(source: bytes, path: str) -> ExtractionResult:
    from tree_sitter import QueryCursor

    parser = _parser()
    query = _query()
    tree = parser.parse(source)
    captures = QueryCursor(query).captures(tree.root_node)

    imports: list[Import] = []
    exports: list[Export] = []

    pending_module: dict[int, str] = {}

    for capture_name, nodes in captures.items():
        for node in nodes:
            text = node.text.decode("utf-8", errors="replace")
            line = node.start_point[0] + 1

            if capture_name == "import.module":
                imports.append(Import(module=text, symbol=None, line=line))

            elif capture_name == "import_from.module":
                pending_module[node.start_byte] = text

            elif capture_name == "import_from.symbol":
                module = _nearest_pending_module(pending_module, node.start_byte)
                imports.append(Import(module=module, symbol=text, line=line))

            elif capture_name == "export.function":
                exports.append(Export(name=text, kind="function", line=line))

            elif capture_name == "export.class":
                exports.append(Export(name=text, kind="class", line=line))

    return ExtractionResult(language="python", imports=imports, exports=exports)


def _nearest_pending_module(pending: dict[int, str], symbol_byte: int) -> str:
    """For `from X import Y, Z`, both Y and Z attach to the same X.

    Tree-sitter captures don't pair them automatically — we look up the
    most recent module capture before the symbol's byte position.
    """
    candidates = [byte for byte in pending if byte < symbol_byte]
    if not candidates:
        return "unknown"
    return pending[max(candidates)]

File: tools/tracer/src/extraction/python.rs
//! Python tree-sitter extraction: module-level imports and definitions.

use crate::extraction::{Export, ExtractionResult, Import};
use std::collections::HashMap;
use tree_sitter::{Parser, Query, QueryCursor, StreamingIterator};

const QUERY_SRC: &str = r#"
        ; import_statement covers `import a`, `import a.b`, `import a as c`
        (import_statement
          name: (dotted_name) @import.module)
        (import_statement
          name: (aliased_import
                  name: (dotted_name) @import.module))

        ; import_from_statement covers `from x import y`, `from x import y as z`
        (import_from_statement
          module_name: (dotted_name) @import_from.module
          name: (dotted_name) @import_from.symbol)
        (import_from_statement
          module_name: (dotted_name) @import_from.module
          name: (aliased_import
                  name: (dotted_name) @import_from.symbol))

        ; module-level definitions only — nested defs are filtered below
        (module
          (function_definition name: (identifier) @export.function))
        (module
          (class_definition name: (identifier) @export.class))
        (module
          (decorated_definition
            definition: (function_definition name: (identifier) @export.function)))
        (module
          (decorated_definition
            definition: (class_definition name: (identifier) @export.class)))
"#;

fn empty() -> ExtractionResult {
    ExtractionResult {
        language: "python".into(),
        imports: vec![],
        exports: vec![],
    }
}

pub fn extract(source: &[u8]) -> ExtractionResult {
    let lang: tree_sitter::Language = tree_sitter_python::LANGUAGE.into();
    let mut parser = Parser::new();
    if parser.set_language(&lang).is_err() {
        return empty();
    }
    let tree = match parser.parse(source, None) {
        Some(t) => t,
        None => return empty(),
    };
    extract_from_tree(&tree, source)
}

/// Imports/exports from an already-parsed Python tree. Decoupled from the
/// parse so `file_facts` shares one tree with `ccn` (single-parse). Caller
/// guarantees `tree` came from the Python grammar.
pub fn extract_from_tree(
    tree: &tree_sitter::Tree,
    source: &[u8],
) -> ExtractionResult {
    let lang: tree_sitter::Language = tree_sitter_python::LANGUAGE.into();
    let query = match Query::new(&lang, QUERY_SRC) {
        Ok(q) => q,
        Err(_) => return empty(),
    };
    let capture_names = query.capture_names();

    #[derive(Clone)]
    struct Cap {
        name: String,
        text: String,
        line: i64,
        start_byte: usize,
    }
    let mut caps: Vec<Cap> = Vec::new();
    let mut cursor = QueryCursor::new();
    let mut it = cursor.matches(&query, tree.root_node(), source);
    while let Some(m) = it.next() {
        for cap in m.captures {
            let node = cap.node;
            caps.push(Cap {
                name: capture_names[cap.index as usize].to_string(),
                text: node.utf8_text(source).unwrap_or("").to_string(),
                line: node.start_position().row as i64 + 1,
                start_byte: node.start_byte(),
            });
        }
    }

    // Group captures by name in first-seen order, then iterate the nodes
    // within each group.
    let mut order: Vec<String> = Vec::new();
    let mut groups: HashMap<String, Vec<Cap>> = HashMap::new();
    for c in caps {
        if !groups.contains_key(&c.name) {
            order.push(c.name.clone());
        }
        groups.entry(c.name.clone()).or_default().push(c);
    }

    let mut imports: Vec<Import> = Vec::new();
    let mut exports: Vec<Export> = Vec::new();

    // Symbol resolution depends only on byte position, not on the order
    // captures are emitted. Collect every pending module first so a
    // `from X import a, b` resolves its symbols correctly regardless of
    // capture-emission order.
    let mut pending_module: Vec<(usize, String)> = Vec::new();
    if let Some(g) = groups.get("import_from.module") {
        for c in g {
            pending_module.push((c.start_byte, c.text.clone()));
        }
    }

    for name in &order {
        for c in &groups[name] {
            match c.name.as_str() {
                "import.module" => imports.push(Import {
                    module: c.text.clone(),
                    symbol: None,
                    line: c.line,
                }),
                "import_from.module" => {}
                "import_from.symbol" => {
                    let module = nearest_pending(&pending_module, c.start_byte);
                    imports.push(Import {
                        module,
                        symbol: Some(c.text.clone()),
                        line: c.line,
                    });
                }
                "export.function" => exports.push(Export {
                    name: c.text.clone(),
                    kind: "function".into(),
                    line: c.line,
                }),
                "export.class" => exports.push(Export {
                    name: c.text.clone(),
                    kind: "class".into(),
                    line: c.line,
                }),
                _ => {}
            }
        }
    }

    ExtractionResult {
        language: "python".into(),
        imports,
        exports,
    }
}

/// The pending `from`-module whose byte offset most closely precedes a
/// symbol — the module that symbol was imported from.
fn nearest_pending(pending: &[(usize, String)], symbol_byte: usize) -> String {
    pending
        .iter()
        .filter(|(b, _)| *b < symbol_byte)
        .max_by_key(|(b, _)| *b)
        .map(|(_, m)| m.clone())
        .unwrap_or_else(|| "unknown".to_string())
}
